File: javascript/01-magic-receipt/test/receipt.test.js
import assert from 'node:assert';
import test from 'node:test';
import { Receipt } from '../src/receipt.js';

test('calculates total for regular customer', () => {
  const receipt = new Receipt();
  assert.strictEqual(receipt.calculate_total([10, 20, 30]), 59.4);
});

test('calculates total for member customer', () => {
  const receipt = new Receipt();
  assert.strictEqual(receipt.calculate_total([10, 20, 30], 'member'), 56.16);
});

test('calculates total for vip customer', () => {
  const receipt = new Receipt();
  assert.strictEqual(receipt.calculate_total([10, 20, 30], 'vip'), 47.68);
});

test('applies bonus discount over 50', () => {
  const receipt = new Receipt();
  assert.strictEqual(receipt.calculate_total([60]), 59.4);
});

test('applies vip extra discount', () => {
  const receipt = new Receipt();
  assert.strictEqual(receipt.calculate_total([100], 'vip'), 84.4);
});

test('returns 0 for empty items', () => {
  const receipt = new Receipt();
  assert.strictEqual(receipt.calculate_total([]), 0);
});

test('exactly 50 gets no bonus discount', () => {
  const receipt = new Receipt();
  assert.strictEqual(receipt.calculate_total([50]), 54);
});

File: csharp/01-magic-receipt/test/receipt_test.cs
using Xunit;

public class ReceiptTest
{
    [Fact]
    public void CalculatesTotalForRegularCustomer()
    {
        var receipt = new Receipt();
        Assert.Equal(59.4m, receipt.calculate_total(new[] { 10m, 20m, 30m }));
    }

    [Fact]
    public void CalculatesTotalForMemberCustomer()
    {
        var receipt = new Receipt();
        Assert.Equal(56.16m, receipt.calculate_total(new[] { 10m, 20m, 30m }, "member"));
    }

    [Fact]
    public void CalculatesTotalForVipCustomer()
    {
        var receipt = new Receipt();
        Assert.Equal(47.68m, receipt.calculate_total(new[] { 10m, 20m, 30m }, "vip"));
    }

    [Fact]
    public void AppliesBonusDiscountOver50()
    {
        var receipt = new Receipt();
        Assert.Equal(59.4m, receipt.calculate_total(new[] { 60m }));
    }

    [Fact]
    public void AppliesVipExtraDiscount()
    {
        var receipt = new Receipt();
        Assert.Equal(84.4m, receipt.calculate_total(new[] { 100m }, "vip"));
    }

    [Fact]
    public void Returns0ForEmptyItems()
    {
        var receipt = new Receipt();
        Assert.Equal(0m, receipt.calculate_total(Array.Empty<decimal>()));
    }

    [Fact]
    public void Exactly50NoBonus()
    {
        var receipt = new Receipt();
        Assert.Equal(54.0m, receipt.calculate_total(new[] { 50m }));
    }
}
